uc-crux-llvm/doc/soundness.md

```diff
@@ -0,0 +1,132 @@
+# Soundness
+
+A program analysis is *sound* if every claim it makes is true. UC-Crux has
+applications as both a bug-finding tool and a verification[^1] tool, and these
+use-cases have different soundness criteria: a bug-finding tool is sound if
+whenever it claims that there is a bug in the program that bug is actually
+possible in some execution of the program; a verification tool is sound if
+whenever it claims that a program is safe every possible execution of the
+program is safe.
+
+Given these complementary properties, there are four possible criteria for
+soundness that each analysis feature of UC-Crux might meet:
+
+- Overapproximate: The feature reflects a superset of the set of possible
+  program behaviors
+- Underapproximate: The feature reflects a subset of the set of possible program
+  behaviors
+- Precise: Both over- and under-approximate, that is, the feature exactly
+  reflects the actual set of possible program behaviors
+- Indefinite: Neither over-approximate nor under-approximate, that is, the
+  feature doesn't necessarily bear any of the above relationships to the set of
+  possible program behaviors; such features might also be called *unsound* to
+  denote that they don't meet a specific soundness criterion
+
+The following table shows how these labels correspond to soundness for both
+verification and bug-finding:
+
+| Property         | Sound for verification | Sound for bugfinding |
+|:-----------------|:----------------------:|:--------------------:|
+| Overapproximate  | Y                      | N                    |
+| Underapproximate | N                      | Y                    |
+| Precise          | Y                      | Y                    |
+| Indefinite       | N                      | N                    |
```

Comment:
I'm definitely open to suggestions - I would like to be as in-line with normal terminology as possible, I think I just don't understand it quite yet. 

One concern I have about using soundness/completeness for bug-finding/verification terminology is that UC-Crux-LLVM doesn't really have the ability to be complete for either:

- For bug-finding, UC-Crux sometimes says "hey, maybe this is a bug or maybe not", and then can't dig deeper into the program, making it incomplete.
- In either case, UC-Crux gives up when e.g. it can't prove loops terminate or it doesn't have an implementation for an external function.

> though you could settle for a partial verification tool (like a semi-decision procedure) that can tell you that the program is correct or that it doesn't know

This would be *incomplete*, correct?

Reply:
> > though you could settle for a partial verification tool (like a semi-decision procedure) that can tell you that the program is correct or that it doesn't know
> 
> This would be _incomplete_, correct?

Yes, I would think of that as sound and incomplete. If it tells you that a program is correct, it is definitely correct. It is not able to identify all correct programs (so is incomplete).